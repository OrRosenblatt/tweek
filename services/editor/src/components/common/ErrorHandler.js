import React, { Component } from 'react';
import PropTypes from 'prop-types';

export default class ErrorHandler extends Component {
  static propTypes = {
    onError: PropTypes.func,
    errorMessage: PropTypes.string,
  };

  state = {
    hasError: false,
  };

  componentDidCatch(error, info) {
    this.setState({ hasError: true });

    const { onError } = this.props;
    onError && onError(error, info);
  }

  render() {
    if (this.state.hasError) {
      const { errorMessage } = this.props;
      return errorMessage ? <div>{errorMessage}</div> : null;
    }

    return this.props.children;
  }
}
