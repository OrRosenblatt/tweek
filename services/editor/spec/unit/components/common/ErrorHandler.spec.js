/* global jest, before, beforeEach, describe, it, expect */
import React from 'react';
import renderer from 'react-test-renderer';
import ErrorHandler from '../../../../src/components/common/ErrorHandler';

const ErrorComponent = () => {
  throw 'error';
};

describe('ErrorHandler component', () => {
  it('should not render when exception is thrown and no error message is passed', () => {
    const tree = renderer.create(
      <ErrorHandler>
        <ErrorComponent />
      </ErrorHandler>,
    );

    expect(tree).toMatchSnapshot();
  });

  it('should render error message when exception is thrown', () => {
    const tree = renderer.create(
      <ErrorHandler errorMessage="some error message">
        <ErrorComponent />
      </ErrorHandler>,
    );

    expect(tree).toMatchSnapshot();
  });

  it('should call onError when exception is thrown', () => {
    const onErrorMock = jest.fn();
    renderer.create(
      <ErrorHandler onError={onErrorMock}>
        <ErrorComponent />
      </ErrorHandler>,
    );

    expect(onErrorMock).toHaveBeenCalled();
  });

  it('should render original component correctly', () => {
    const onErrorMock = jest.fn();

    const tree = renderer.create(
      <ErrorHandler errorMessage="some error message" onError={onErrorMock}>
        <div>some element</div>
        <div>another element</div>
      </ErrorHandler>,
    );

    expect(onErrorMock).not.toHaveBeenCalled();
    expect(tree).toMatchSnapshot();
  });
});
