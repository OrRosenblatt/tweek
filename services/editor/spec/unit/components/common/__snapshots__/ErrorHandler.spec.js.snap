// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ErrorHandler component should not render when exception is thrown and no error message is passed 1`] = `null`;

exports[`ErrorHandler component should render error message when exception is thrown 1`] = `
<div>
  some error message
</div>
`;

exports[`ErrorHandler component should render original component correctly 1`] = `
Array [
  <div>
    some element
  </div>,
  <div>
    another element
  </div>,
]
`;
